### ISSUE TITLE
ttnn.sparse_matmul deadlocks when declared nnz > count_nonzero(sparsity)

### ISSUE THREAD

**Issue by djordje-tt** [tenstorrent/tt-metal#45943]
# `ttnn.sparse_matmul` deadlocks when declared `nnz` > actual `count_nonzero(sparsity)`

## Summary
`ttnn.sparse_matmul` (the 1D in0-mcast program) **hangs the device** when the caller passes an `nnz` value that is **larger than the number of non-zero entries in the `sparsity` tensor at runtime**. The op silently assumes `count_nonzero(sparsity) == nnz`; when that invariant is violated (actual < declared), the in0-multicast receivers wait forever on a multicast the sender never issues, the op never completes, and the device wedges (`TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable`).

This is the op-level root cause of the **gpt-oss-120B Blackhole decode hang (#45052)**. (See "How it bit gpt-oss" below.)

## Repro
Branch: **`sparse-matmul-nnz-mismatch-hang`**
Test: `tests/ttnn/unit_tests/operations/matmul/test_sparse_matmul_nnz_mismatch_repro.py`

Single device, no mesh / no fabric / no model. The two tests differ **only** in the `nnz` value passed.

```bash
# Control — declared nnz == actual non-zeros -> PASSES (~4.5s)
pytest tests/ttnn/unit_tests/operations/matmul/test_sparse_matmul_nnz_mismatch_repro.py::test_nnz_matches_actual_passes -x -v

# Bug — declared nnz (4) > actual non-zeros (1) -> HANGS (deadlock)
TT_METAL_OPERATION_TIMEOUT_SECONDS=30 \
  pytest tests/ttnn/unit_tests/operations/matmul/test_sparse_matmul_nnz_mismatch_repro.py::test_nnz_greater_than_actual_hangs -x -v
```

Verified on a single Blackhole card, built from `main`:

| Test | `nnz` passed | actual non-zeros | Result |
|---|---|---|---|
| `test_nnz_matches_actual_passes` | 1 | 1 | ✅ `1 passed in 4.50s` |
| `test_nnz_greater_than_actual_hangs` | 4 | 1 | ❌ deadlock → device timeout after 30s |

Bug-case log:
```
declared nnz=4, actual count_nonzero(sparsity)=1
... 30s later ...
Metal | Timeout detected (metal_context.cpp:752)
Always | TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable
```

The defect is a control-flow/semaphore mismatch and is **architecture-agnostic** (reproduces on WH and BH); Blackhole is only where the model happens to violate the invariant (see below).

## Root cause (in0-mcast handshake)
When `nnz` is supplied, the factory (`sparse_matmul_multicore_reuse_mcast_1d_optimized.cpp`) sets `get_batch_from_reader = !nnz.has_value() = false` and wires the kernels with **inconsistent loop counts**:

| kernel | loop count | reads `sparsity`? |
|---|---|---|
| in0 **sender** (`reader_bmm_tile_layout_in0_sender_padding.cpp`) | `batchB` (all experts) | yes — **`continue`s (no multicast) on every zero entry** |
| in1 reader (`reader_bmm_tile_layout_in1_sender_writer_padding.cpp`) | `batchB` | yes — skips zeros |
| in0 **receiver** (`reader_bmm_tile_layout_in0_receiver.cpp`) | `num_batch_compute = nnz` | no |
| compute (`bmm_large_block_zm_fused_bias_activation.cpp`) | `nnz` | no |

So the **sender multicasts exactly `count_nonzero(sparsity)` times**, while the **receivers/compute loop a fixed `nnz` times**, each iteration waiting on the in0 semaphore.

If `count_nonzero(sparsity) < nnz`:
1. The sender finishes its `batchB` scan having multicast only `count_nonzero` blocks, then its kernel exits.
2. Receiver iteration `count_nonzero` (and beyond) does `sender_sem.up()` then `receiver_sem.wait(VALID)` — but the sender has already exited and will never set `VALID` again → **receiver blocks forever in `noc_semaphore_wait`**.
3. Compute blocks in `llk_wait_tiles` waiting for in0 blocks that never arrive; the op never retires; the device wedges.

(`count_nonzero > nnz` is also unsafe — the sender over-produces and back-pressures — but the deadlock observed in the field and in the repro is the `actual < nnz` direction.)

The op contract (`matmul_nanobind.cpp`) documents `nnz` as optional — *"If not provided, it will be inferred from the sparsity tensor at runtime"* — and `0 ≤ nnz ≤ E`. But there is no validation that `nnz == count_nonzero(sparsity)`, and a mismatch deadlocks instead of erroring.

## How it bit gpt-oss (#45052)
gpt-oss-120B decode calls `sparse_matmul(..., nnz=num_experts_per_tok=4)` for the MoE gate/up/down projections (`models/demos/gpt_oss/tt/experts/decode.py`). The decode `sparsity` is built from the router as `scatter(zeros, top4_idx, softmax(top4_logits))` (`topk.py`). On **Blackhole**, the softmax over the top-4 logits flushes the smaller weights to **exact 0.0**, so for many tokens the scattered sparsity has **fewer than 4 non-zeros**. Measured over 6508 decode sparse_matmul calls: count was **1 (60%), 2 (3.7%), 3 (0.6%), 4 (36%)** — i.e. `nnz=4` is wrong ~64% of the time, and any token with count<4 deadlocks. (Wormhole keeps those small weights non-zero, so the count stays 4 there and the hang does not reproduce — which is why #45052 regressed only when Blackhole was enabled.)

tt-triage on the hung gpt-oss run shows exactly this signature: `SparseMatmulDeviceOperation` RUNNING on all devices; `reader_bmm_tile_layout_in0_receiver` parked in `noc_semaphore_wait`; compute in `llk_wait_tiles`; the in0 sender core's BRISC finished and sitting in firmware `wait_ncrisc_trisc`.

## Suggested fix (op owner to decide)
The op should not deadlock on a stale/incorrect `nnz`. Options:
1. **Validate**: `TT_FATAL` (or graceful error) if a supplied `nnz` does not match `count_nonzero(sparsity)` — at least fail loudly instead of hanging. (Note: `count_nonzero` is data-dependent/runtime, so validation may need to be device-side or documented as caller responsibility.)
2. **Be robust**: drive the in0 receiver/compute loop counts from the actual sparsity (the existing `get_batch_from_reader` path) even when `nnz` is supplied, treating `nnz` only as an upper bound for sizing. (Tradeoff: loses the fixed-count fast path; ~13–16% slower in the gpt-oss decode case because it adds a per-expert mcast handshake for skipped experts.)

Model-side workarounds exist (drop `nnz`, or floor the router weights so `count==nnz`), but the **op deadlocking on a count mismatch is the underlying defect**.

## Environment
- Blackhole (also expected on Wormhole); single device.
- `main` @ `dc7b33cdab7` (0.72.0-dev).

Related: #45052

**Comment by bbradelTT** [comment]
@djordje-tt I'll go with option 1, validate.

@amorrisonTT @gajanan-choudhary can what is using sparse_matmul be migrated to generalized moe?

**Comment by bbradelTT** [comment]
@djordje-tt I have the validation on bbradel-45943_sparse_mm_check / draft PR https://github.com/tenstorrent/tt-metal/pull/45958 
Please test it out and let me know if that meets your needs.

**Comment by djordje-tt** [comment]
This works for me since unit test is failing on the check. Unfortunately this requires watcher because it's validated in runtime, right?

**Comment by bbradelTT** [comment]
> This works for me since unit test is failing on the check. Unfortunately this requires watcher because it's validated in runtime, right?

Yes, this requires watcher. Without that, you're paying a perf penalty that we don't want to impose. That's why there are two walls of text in the documentation.

**Comment by bbradelTT** [comment]
@djordje-tt the PR got merged. Are you okay to close this issue?

**Comment by djordje-tt** [comment]
Yes!
